### ISSUE TITLE
`dataclass_transform` on subclasses loses parent class fields

### ISSUE THREAD

**Issue by igor47** [microsoft/pyright#9782]
using pyright in neovim via mason, my pyright version is `1.1.388`

```python
@dataclass_transform()
class Parent():
    parent: str

@dataclass_transform()
class Child(Parent):
    child: str

c = Child()
```

the signature of `Child` is `Child(child: str)` -- what happened to the `parent` field?

I saw (in #3907) that i can generally use `@dataclass` to understand how `@dataclass_transform` should work. trying this:

```python
from dataclasses import dataclass

@dataclass
class Parent():
    parent: str

@dataclass
class Child(Parent):
    child: str

c = Child()
```

results in the expected behavior, with the signature of `Child()` now `Child(parent: str, child: str)`.

**Comment by erictraut** [comment]
You're using `@dataclass_transform` incorrectly. This decorator should be applied only to the base class or metaclass that implements the dataclass-like behavior. You're indicating in this code sample that both `Parent` and `Child` implement dataclass-like behavior. The type checker behavior in this case is undefined.

Here's how this would normally be done:

```python
@dataclass_transform()
class ModelBase: ...

class Parent(ModelBase):
    parent: str

class Child(Parent):
    child: str
```

Or if your dataclass-like behavior is implemented in a metaclass:

```python
@dataclass_transform()
class ModelMeta(type): ...

class Parent(metaclass=ModelMeta):
    parent: str

class Child(Parent):
    child: str
```

For more details about `dataclass_transform`, refer to [this section](https://typing.readthedocs.io/en/latest/spec/dataclasses.html#the-dataclass-transform-decorator) of the Python typing spec.
